### DISCUSSION TITLE
Frontend JS Organization

### DISCUSSION THREAD

**Discussion by justinbeaty** [MahoCommerce/maho#84] [Ideas]
Let's start a discussion about how to ultimately organize the js directory. Here's one possible idea. 

Say we organize like this:

```
js
├── lib
│   ├── flatpickr
│   │   ├── flatpickr.min.css
│   │   └── flatpickr.min.js
│   └── sortable.min.js
├── mage
│   ├── adminhtml
│   │   ├── catalog
│   │   │   ├── category.js
│   │   │   ├── product
│   │   │   │   └── configurable.js
│   │   │   └── product.js
│   │   ├── compat.js
│   │   └── eav
│   │       ├── attribute.js
│   │       └── set.js
│   └── frontend
│       ├── checkout
│       │   └── opcheckout.js
│       ├── compat.js
│       └── customer
│           └── account.js
├── maho
│   ├── autocomplete.js
│   ├── dialog.js
│   └── tree.js
└── util
    ├── compat.js
    ├── cookies.js
    ├── dom.js
    └── rules.js

```

The main directories are:

* `js/lib` - 3rd party libraries
* `js/mage/adminhtml` - backend specific classes.
* `js/mage/frontend` - frontend specific classes.
* `js/maho` - new libraries we create.
* `js/util` - misc utility functions that can be used for either frontend / backend.

--- 

To avoid namespace collisions, each file could define it's classes / methods like the below example. For example:

In `mage/adminhtml/catalog/category.js`:

```js
/**
 * Maho
 *
 * @category    Mage
 * @package     Mage_Adminhtml
 * @copyright   Copyright (c) 2024 Maho (https://mahocommerce.com)
 * @license     https://opensource.org/licenses/afl-3.0.php  Academic Free License (AFL 3.0)
 */

Mage ??= {};
Mage.Adminhtml ??= {};
Mage.Adminhtml.Catalog ??= {};

Mage.Adminhtml.Catalog.Category = class {
    constructor() {
        // ...
    }
};

// In some other file:
new Mage.Adminhtml.Catalog.Category();
```

In `util/dom.js`:

```js
/**
 * Maho
 *
 * @category    Mage
 * @package     Mage_Util
 * @copyright   Copyright (c) 2024 Maho (https://mahocommerce.com)
 * @license     https://opensource.org/licenses/afl-3.0.php  Academic Free License (AFL 3.0)
 */

Mage ??= {};
Mage.Util ??= {};
Mage.Util.Dom ??= {};


Mage.Util.Dom.setMessagesDiv = function () {

}


Mage.Util.Dom.popWin = function () {
}
```
---

However, since that second example is totally backwards incompatible with code that expects `popWin` in the global namespace, we can have a few `compat.js` files.

* `mage/adminhtml/compat.js` - loaded on the backend
* `mage/frontend/compat.js` - loaded on the frontend
* `util/compat.js` - loaded on both

The last one might look something like this, so if `util/dom.js` is loaded, then `window.popWin()` will be defined.

```js
const popWin = Mage?.Util?.Dom?.popWin;
const formatCurrency = Mage?.Util?.formatCurrency;
```

Instead of `compat.js`, `util/dom.js` could also have just done this:

```js
window.popWin = Mage.Util.Dom.popWin = function () {
}
```

Or:
```js
Mage.Util.Dom.popWin = function () {
}
Mage.Util.Dom.otherFn = function () {
}


// at the end of the file:
window.popWin = Mage.Util.Dom.popWin;
window.otherFn = Mage.Util.DomotherFn;
```

The latter ideas are easier, but the former might give more control to a store owner who wants to disable `compat.js` loading to not pollute the global namespace.

---

In a perfect world, we could be using es6 modules, but I don't think we're anywhere near that. This, or some other idea, would still be a step in the right direction.

The above is also just an example, we don't need to go crazy with splitting code into so many small chunks. So maybe the `util` dir doesn't need to have `util/dom.js`, `util/fetch.js`, `util/messages.js`, etc. Maybe somewhere in between fully modular and the massive `js.js` file we have now.

**Comment by justinbeaty** [comment]
It might need to be:

```js
var Mage = Mage ?? {};
Mage.Adminhtml ??= {};
Mage.Adminhtml.Catalog ??= {};
```

Or:

```js
window.Mage ??= {};
Mage.Adminhtml ??= {};
Mage.Adminhtml.Catalog ??= {};
```

I created a [test repo](https://github.com/justinbeaty/js-org-test/tree/main). Does this work with your IDE's autocomplete? There's only a few things defined:

```js
Mage.Adminhtml.Catalog.Category
Mage.Util.Dom.setMessagesDiv
Mage.Util.Dom.popWin

setMessagesDiv
popWin
```

Edit: I tested in PHPStorm and it seems to work fine.

**Comment by fballiano** [comment]
sorry man, I'll write soon :-) thanks for this in the meanwhile!

**Reply by justinbeaty** [reply]
No problem, have been busy all day today anyway.

**Comment by fballiano** [comment]
> 
> The main directories are:
> 
> * `js/lib` - 3rd party libraries
> * `js/mage/adminhtml` - backend specific classes.
> * `js/mage/frontend` - frontend specific classes.
> * `js/maho` - new libraries we create.
> * `js/util` - misc utility functions that can be used for either frontend / backend.

I was writing a longer answer with some considerations but at the end of the day I agree with your proposal 100% so.. :-D

> In `mage/adminhtml/catalog/category.js`:
> 
> ```js
> /**
>  * Maho
>  *
>  * @category    Mage
>  * @package     Mage_Adminhtml
>  * @copyright   Copyright (c) 2024 Maho (https://mahocommerce.com)
>  * @license     https://opensource.org/licenses/afl-3.0.php  Academic Free License (AFL 3.0)
>  */
> 
> Mage ??= {};
> Mage.Adminhtml ??= {};
> Mage.Adminhtml.Catalog ??= {};
> 
> Mage.Adminhtml.Catalog.Category = class {
>     constructor() {
>         // ...
>     }
> };
> 
> // In some other file:
> new Mage.Adminhtml.Catalog.Category();
> ```

yes, since probably a real class is needed anyway there


> In `util/dom.js`:
> 
> ```js
> /**
>  * Maho
>  *
>  * @category    Mage
>  * @package     Mage_Util
>  * @copyright   Copyright (c) 2024 Maho (https://mahocommerce.com)
>  * @license     https://opensource.org/licenses/afl-3.0.php  Academic Free License (AFL 3.0)
>  */
> 
> Mage ??= {};
> Mage.Util ??= {};
> Mage.Util.Dom ??= {};
> 
> 
> Mage.Util.Dom.setMessagesDiv = function () {
> 
> }
> 
> 
> Mage.Util.Dom.popWin = function () {
> }
> ```

yes, since utilities probably are just a bunch of functions and don't need a real class that the use would have to instantiate

> * `util/compat.js` - loaded on both

I this I'd like to have a compat.js file so that it can be excluded by the store developer, but also it's easier for us to just remove it at some point. Maybe I'd just have one single compat file instead of 3, since it will be just a few lines of code.

but loading order could be a problem in some cases (when 3rd party loads a js in some weird way)... mmm...

> The above is also just an example, we don't need to go crazy with splitting code into so many small chunks. So maybe the `util` dir doesn't need to have `util/dom.js`, `util/fetch.js`, `util/messages.js`, etc. Maybe somewhere in between fully modular and the massive `js.js` file we have now.

nah I wouldn't like to have many different files, I think the way we described above it's perfectly fine :-)

**Reply by justinbeaty** [reply]
> I this I'd like to have a compat.js file so that it can be excluded by the store developer, but also it's easier for us to just remove it at some point. Maybe I'd just have one single compat file instead of 3, since it will be just a few lines of code.

One file sounds good to me.

> but loading order could be a problem in some cases (when 3rd party loads a js in some weird way)... mmm...

This should work, I tested by loading this file before dom.js:

```js
const popWin = (...args) => Mage?.Util?.Dom?.popWin(...args);
const setMessagesDiv = (...args) => Mage?.Util?.Dom?.setMessagesDiv(...args);
```

> nah I wouldn't like to have many different files, I think the way we described above it's perfectly fine :-)

Agreed, we definitely don't want super specific files with only like 2-3 functions defined in each.

**Comment by justinbeaty** [comment]
A couple more examples:

Depending on the file it might be cleaner to define like this, as long as no other file also puts members into the same namespace:

```js
Mage.Util.Dom = {

    setMessagesDiv(message, type = 'success') {
        console.log('Called Mage.Util.Dom.setMessagesDiv()');
    },

    popWin() {
        console.log('Called Mage.Util.Dom.popWin()');
    },
};
```

---

Creating a subclass:

```js
Mage.Adminhtml.Catalog.Category = class {
    constructor(config) {
        console.log('Called new Mage.Adminhtml.Catalog.Category()');
    }
    foo() {
        console.log('Category.foo()');
    }
};

Mage.Adminhtml.Catalog.SubCategory = class extends Mage.Adminhtml.Catalog.Category {
    constructor() {
        super();
        console.log('Called new Mage.Adminhtml.Catalog.SubCategory()');
    }
    foo() {
        console.log('SubCategory.foo()');
    }
}
```

**Reply by fballiano** [reply]
for proper classes I think this method is ok (EG: configurable.js) but for utils I think `Mage.Util.setMessagesDiv = function...` seems more flexible to me

**Comment by justinbeaty** [comment]
Few other points:

1\. I think the util folder should actually be `public/js/mage/util`, so that we have `Mage.Util.Whatever` instead of just `Util.Whatever`. I suppose it's okay that we put new functions there instead of having both `Mage.Util` and `Maho.Util` ?

2\. For existing custom modules and themes with a layout xml file, do we need to provide any sort of mapping from old files to new? For example, a module wants to load `varien/accordion.js` but we've moved it to `mage/util/accordion.js`. Do we define aliases in `Mage_Page_Block_Html_Head`?

**Reply by fballiano** [reply]
> 1. I think the util folder should actually be `public/js/mage/util`, so that we have `Mage.Util.Whatever` instead of just `Util.Whatever`. I suppose it's okay that we put new functions there instead of having both `Mage.Util` and `Maho.Util` ?

mmmm what if we remove the distinction between Mage and Maho and just migrate everything under the Maho namespace providing compatibility via the compat file? cause having js/mage and js/maho seems ok from our perspective but the end user it's not that great

> 2. For existing custom modules and themes with a layout xml file, do we need to provide any sort of mapping from old files to new? For example, a module wants to load `varien/accordion.js` but we've moved it to `mage/util/accordion.js`. Do we define aliases in `Mage_Page_Block_Html_Head`?

mmm this is though, does it make sense to do that? will it just be a comment in a layout.xml that nobody would see? or should we document it in the release notes and that's it?

**Comment by justinbeaty** [comment]
> and just migrate everything under the Maho namespace

Yes, that sounds good to me!

> mmm this is though, does it make sense to do that? will it just be a comment in a layout.xml that nobody would see? or should we document it in the release notes and that's it?

I was thinking something more like this in `app/code/core/Mage/Page/Block/Html/Head.php`:

```php
public function addItem($type, $name, $params = null, $if = null, $cond = null, $referenceName = '*', $before = false)
{
    if ($type === 'js') {
        $name = match ($name) {
            'varien/accordion.js' => 'maho/util/accordion.js',
            'varien/form.js' => 'maho/util/form.js',
            // ...
            default => $name,
        };
    }
```

What do you think about making a configuration option like "JavaScript compatibility mode" that controls both including `compat.js` and enables the mapping above? The downside is that it's another config option, but it would allow us to ensure `compat.js` is always loaded first, and allows store owners to disable it without having to modify both the frontend and backend theme layout files.

**Reply by fballiano** [reply]
>     if ($type === 'js') {
>         $name = match ($name) {
>             'varien/accordion.js' => 'maho/util/accordion.js',
>             'varien/form.js' => 'maho/util/form.js',
>             // ...
>             default => $name,
>         };
>     }
> ```

cause you're a genius, I wouldn't have thought about that 

> What do you think about making a configuration option like "JavaScript compatibility mode" that controls both including `compat.js` and enables the mapping above?

I love that idea actually!

**Comment by fballiano** [comment]
Something not to forget: https://github.com/MahoCommerce/maho/pull/136#discussion_r1999758329
